### DISCUSSION TITLE
How to tell ML to avoid previously labeled segmentations?

### DISCUSSION THREAD

**Discussion by ross-mitchell** [Project-MONAI/MONAILabel#1145] [General]
Hello:

I am segmenting several objects in the same scene (multiple organs in abdominal CT scans). There are situations where it would be very helpful to tell MONAI Label to avoid already labeled structures (besides background). That is, tell ML not to grow into voxel regions that already have a non-background label.

Is this easily doable?

**Comment by danieltudosiu** [comment]
Hi Ross,

To do that you would need to integrate some sort of partial update of the
segmentations. At the top of my mind come those papers that might be of
interest:

Lin, Z., Duan, Z.P., Zhang, Z., Guo, C.L. and Cheng, M.M., 2022. FocusCut:
Diving Into a Focus View in Interactive Segmentation. In *Proceedings of
the IEEE/CVF Conference on Computer Vision and Pattern Recognition* (pp.
2637-2646).

Chen, X., Zhao, Z., Zhang, Y., Duan, M., Qi, D. and Zhao, H., 2022.
FocalClick: Towards Practical Interactive Image Segmentation. In *Proceedings
of the IEEE/CVF Conference on Computer Vision and Pattern Recognition* (pp.
1300-1309).

Sofiiuk, K., Petrov, I., Barinova, O. and Konushin, A., 2020. f-brs:
Rethinking backpropagating refinement for interactive segmentation. In
*Proceedings
of the IEEE/CVF Conference on Computer Vision and Pattern Recognition* (pp.
8623-8632).

About the ease of implementation, it all depends on how familiar you are
with MONAI and Ignite libraries, and to some degree with PyTorch Lightning
due to similarities to MONAI Label.

From my skimming of the papers, it would require some sort of data cropping
based on preexisting segmentations. Sadly I am not that familiar with
Datastore and how MONAI Label handles data processing to answer you in
detail.

Hope it helps,

Dan

**Comment by diazandr3s** [comment]
> Hello:
> 
> I am segmenting several objects in the same scene (multiple organs in abdominal CT scans). There are situations where it would be very helpful to tell MONAI Label to avoid already labeled structures (besides background). That is, tell ML not to grow into voxel regions that already have a non-background label.
> 
> Is this easily doable?

Hi @ross-mitchell,

It is not easy to tell the MONAI Label model to not grow into custom regions. It'll always get the predictions for all the labels it was trained on. Unless you want to implement what @danieltudosiu recommends.
You could get all the predictions from the pretrained and then modify them in the viewer - all this is more of a viewer operation I'd say.

Hope this makes sense. Happy to discuss this further in a videocall if needed.
